db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ExcludedIndexColumnVisitor.java

```diff
@@ -0,0 +1,119 @@
+ * the Apache License, Version 2.0 (the "License"); you may not use these files
+ * except in compliance with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Splice Machine, Inc. has modified the Apache Derby code in this file.
+ *
+ * All such Splice Machine modifications are Copyright 2012 - 2017 Splice Machine, Inc.,
+ * and are licensed to you under the GNU Affero General Public License.
+ */
+
+package com.splicemachine.db.impl.sql.compile;
+
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.sql.compile.C_NodeTypes;
+import com.splicemachine.db.iapi.sql.compile.Visitable;
+import com.splicemachine.db.iapi.sql.compile.Visitor;
+import com.splicemachine.db.iapi.types.DataValueDescriptor;
+
+
+/**
+ * This visitor probes the predicates looking for restrictions on the first column in a key.
+ *
+ */
+public class ExcludedIndexColumnVisitor implements Visitor {
+    int tableNumber;
+    int columnNumber;
+    boolean nullsExcluded;
+    boolean isValid;
+    DataValueDescriptor defaultValue;
+
+
+    public ExcludedIndexColumnVisitor(int tableNumber, int columnNumber, boolean nullsExcluded, DataValueDescriptor defaultValue){
+        this.tableNumber = tableNumber;
+        this.columnNumber = columnNumber;
+        this.isValid = false;
+        this.nullsExcluded = nullsExcluded;
+        this.defaultValue = defaultValue;
+    }
+
+
+    public Visitable visit(Visitable node, QueryTreeNode parent) throws StandardException {
+        if(node instanceof ColumnReference){
+            ColumnReference cr = (ColumnReference) node;
+
+            if(cr.getSource() != null &&
+                cr.getTableNumber() == tableNumber &&
+                cr.getColumnNumber() == columnNumber) {
+                if (parent != null &&
+                        parent.getNodeType() == C_NodeTypes.IS_NULL_NODE &&
+                        nullsExcluded) {
+                    // Predicate does not validate a restriction on the index.
+                } else if (parent != null &&
+                        parent instanceof BinaryRelationalOperatorNode && // BRO
+                        ((BinaryRelationalOperatorNode) parent).getRightOperand() instanceof ConstantNode) {
+                    DataValueDescriptor compare = ((ConstantNode) ((BinaryRelationalOperatorNode) parent).getRightOperand()).getValue();
+                    if (compare != null && compare.equals(defaultValue)) {
+                        // Predicate Does not restrict since it is doing a predicate comparison
```

Comment:
The check here seems to be too loose. If we have binary relation operator other than equality, checking whether the right constant value is equal to the default value is not enough. For example:
create table food4 (col1 int, col2 int default  5);
create index food4_ix on food4(col2) exclude default keys;
insert into food4(col1) values 1,2,3;

The following query will return wrong result:
select * from food4 f --splice-properties index=food4_ix
where col2 > 3;